File: fsensor/src/main/java/com/tracqi/fsensor/math/offset/Calibration.java
package com.tracqi.fsensor.math.offset;

import org.apache.commons.math3.linear.RealMatrix;
import org.apache.commons.math3.linear.RealVector;

/*
 * Copyright 2024, Tracqi Technology, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Represents a field calibration.
 * Created by kaleb on 3/18/18.
 */
public class Calibration {
    public final RealMatrix scalar;
    public final RealVector offset;

    /**
     * Create an instance.
     *
     * @param scalar The scalar of the calibration.
     * @param offset The offset of the calibration.
     */
    public Calibration(RealMatrix scalar, RealVector offset) {
        this.scalar = scalar;
        this.offset = offset;
    }

}

File: fsensor/src/commonMain/kotlin/com/tracqi/fsensor/math/offset/Calibration.kt
package com.tracqi.fsensor.math.offset

data class Calibration(val scalar: Array<DoubleArray>, val offset: DoubleArray) {
    fun calibrate(vector: FloatArray): FloatArray {
        val x = vector[0].toDouble() - offset[0]
        val y = vector[1].toDouble() - offset[1]
        val z = vector[2].toDouble() - offset[2]
        return floatArrayOf(
            (scalar[0][0] * x + scalar[0][1] * y + scalar[0][2] * z).toFloat(),
            (scalar[1][0] * x + scalar[1][1] * y + scalar[1][2] * z).toFloat(),
            (scalar[2][0] * x + scalar[2][1] * y + scalar[2][2] * z).toFloat()
        )
    }
}
